tests/parser/WellTracerTests.cpp

```diff
@@ -0,0 +1,121 @@
+*/
+
+#define BOOST_TEST_MODULE WellTracerTests
+
+#include <boost/filesystem.hpp>
+#include <boost/test/unit_test.hpp>
+
+#include <opm/parser/eclipse/EclipseState/Grid/EclipseGrid.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/Schedule.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/Well.hpp>
+#include <opm/parser/eclipse/Deck/Deck.hpp>
+#include <opm/parser/eclipse/Deck/DeckItem.hpp>
+#include <opm/parser/eclipse/Deck/DeckKeyword.hpp>
+#include <opm/parser/eclipse/Deck/DeckRecord.hpp>
+#include <opm/parser/eclipse/Parser/Parser.hpp>
+#include <opm/parser/eclipse/Parser/ParseContext.hpp>
+
+using namespace Opm;
+
+static Deck createDeckWithOutTracer() {
+    Opm::Parser parser;
+    std::string input =
+            "SCHEDULE\n"
+            "WELSPECS\n"
+            "     'W_1'        'OP'   2   2  1*       \'OIL\'  7* /   \n"
+            "/\n"
+            "COMPDAT\n"
+            " 'W_1'  2*  1   1 'OPEN' / \n"
+            "/\n"
+            "WCONINJE\n"
+            "     'W_1' 'WATER' 'OPEN' 'BHP' 1 2 3/\n/\n";
+
+    return parser.parseString(input, ParseContext());
+}
+
+
+static Deck createDeckWithDynamicWTRACER() {
+    Opm::Parser parser;
+    std::string input =
+            "START             -- 0 \n"
+            "1 JAN 2000 / \n"
+            "SCHEDULE\n"
+            "WELSPECS\n"
+            "     'W_1'        'OP'   1   1  1*       \'GAS\'  7* /   \n"
+            "/\n"
+            "COMPDAT\n"
+            " 'W_1'  2*  1   1 'OPEN' / \n"
+            "/\n"
+            "WCONINJE\n"
+            "     'W_1' 'GAS' 'OPEN' 'BHP' 1 2 3/\n/\n"
+            "DATES             -- 1\n"
+            " 1  MAY 2000 / \n"
+            "/\n"
+            "WTRACER\n"
+            "     'W_1' 'I1'       1 / \n "
+            "     'W_1' 'I2'       1 / \n "
+            "/\n"
+            "DATES             -- 2, 3\n"
+            " 1  JUL 2000 / \n"
+            " 1  AUG 2000 / \n"
+            "/\n"
+            "WTRACER\n"
+            "     'W_1' 'I1'       0 / \n "
+            "/\n"
+            "DATES             -- 4\n"
+            " 1  SEP 2000 / \n"
+            "/\n";
+
+    return parser.parseString(input, ParseContext());
+}
+
+
+
+BOOST_AUTO_TEST_CASE(TestNoTracer) {
+    auto deck = createDeckWithOutTracer();
+    EclipseGrid grid(10,10,10);
+    TableManager table ( deck );
+    Eclipse3DProperties eclipseProperties ( deck , table, grid);
+    Schedule schedule(deck, grid , eclipseProperties, Phases(true, true, true) , ParseContext());
+    BOOST_CHECK(!deck.hasKeyword("WTRACER"));
+}
+
+
+BOOST_AUTO_TEST_CASE(TestDynamicWTRACER) {
+    auto deck = createDeckWithDynamicWTRACER();
+    EclipseGrid grid(10,10,10);
+    TableManager table ( deck );
+    Eclipse3DProperties eclipseProperties ( deck , table, grid);
+    Schedule schedule(deck, grid , eclipseProperties, Phases(true, true, true) , ParseContext());
+    BOOST_CHECK(deck.hasKeyword("WTRACER"));
```

Comment:
Very good that you have written tests - but I miss testing of the sorry-path; as I see it that is actually the most important testing. You have introduced a new `throw ...` statement for the situation where you try to update the tracer properties of a producer, that should be tested with something like:
```
auto deck = createDeckWithTracerInProducer();
...
BOOST_REQUIRE_THROW( Schedule(deck, grid, ecl, ..), std::invalid_argument);
```

Reply:
Thank you, this is now done.